numba/tests/test_dispatcher.py

```diff
@@ -1641,6 +1641,53 @@ def cache_file_collision_tester(q, tempdir, modname_bar1, modname_bar2):
     q.put(r2)
 
 
+class TestCacheMultipleFilesWithSignature(unittest.TestCase):
+    # Regression test for https://github.com/numba/numba/issues/3658
+
+    _numba_parallel_test_ = False
+
+    source_text_file1 = """
+from file2 import function2
+"""
+    source_text_file2 = """
+from numba import njit
+
+@njit('float64(float64)', cache=True)
+def function1(x):
+    return x
+
+@njit('float64(float64)', cache=True)
+def function2(x):
+    return x
+"""
+
+    def setUp(self):
+        self.tempdir = temp_directory('test_cache_file_loc')
```

Comment:
Could this perhaps be removed in a `tearDown()` function?

Reply:
Just pushed a commit to do this (also includes a few other instances of the same pattern occurring in `test_dispatcher.py`). 

I originally thought that `temp_directory` would handle removing directories it creates automatically, but it looks like instead it [periodically removes directories once they become stale](https://github.com/numba/numba/blob/86f98326a58d356fa8e346a1698d47a6e3869729/numba/tests/support.py#L518-L520). Would it be worth opening an issue to go through the test suite and add `tearDown()` methods to remove directories created by `temp_directory`?

Reply:
The "periodic cleanup" was done that way so that the temp directory can be inspected after the testsuite is done.  We should revisit its design.

Reply:
Added issue https://github.com/numba/numba/issues/4274

Reply:
Good to know, thanks for opening up the issue @sklam : )